node/src/components/consensus/era_supervisor.rs

```diff
@@ -145,9 +145,21 @@ where
         let metrics = ConsensusMetrics::new(registry)
             .expect("failure to setup and register ConsensusMetrics");
         let genesis_start_time = protocol_config.timestamp;
-        let current_era = protocol_config.last_upgrade_point;
 
-        let mut era_supervisor = Self {
+        let effects = effect_builder
+            .collect_switch_blocks(
+                current_era.iter_other(protocol_config.last_upgrade_point, bonded_eras * 3),
+            )
+            .event(move |switch_blocks| Event::InitializeEras {
+                switch_blocks: switch_blocks.expect("should have all the switch blocks in storage"),
```

Comment:
That's not enough. We should verify that we have _all_ the switch blocks – i.e. that the number of switch blocks in `switch_block` is as expected.

Reply:
If we don't have all, this will be `None`. It's ensured by the effect (if any of the reads returns `None`, the effect as a whole returns `None`). Maybe it's too much of a logic split, though - maybe all of it should be here, or all of it in the effect?

Reply:
Is that what `join_all!` does?

Reply:
No - it's done by `into_iter().collect()`. `join_all` will return a `Vec<Option<...>>`. Iterating through this and collecting into `Option<HashMap<...>>` produces `None` if any of the elements in the iterator are `None`. This is because of the way `FromIterator` is implemented for `Option`s.

Reply:
OK, thanks :+1: